### ISSUE TITLE
train_hmm.py

### ISSUE THREAD

**Issue by ZhigangSu** [ZhigangSu/HMM#1]
import numpy as np
import os
import pickle
import argparse
from sklearn.datasets import fetch_openml
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix
import matplotlib.pyplot as plt
from hmmlearn import hmm
from PIL import Image

def extract_features(images):
    h, w = images.shape[1], images.shape[2]
    seqs = []
    for img in images:
        bin_img = (img > 128).astype(int)
        dens = bin_img.sum(axis=0) / h
        seqs.append(dens.reshape(-1,1))
    return seqs

if __name__=='__main__':
    parser = argparse.ArgumentParser(description="Train HMM for MNIST digits")
    parser.add_argument('--train_samples', type=int, default=1000)
    parser.add_argument('--states', type=int, default=5)
    parser.add_argument('--iter', type=int, default=100)
    parser.add_argument('--model_dir', type=str, default='models')
    parser.add_argument('--test_custom', type=str, default=None, help='测试单张自定义图像')
    parser.add_argument('--test_folder', type=str, default=None, help='测试图像目录')
    args = parser.parse_args()

    if not args.test_custom and not args.test_folder:
        
        mnist = fetch_openml('mnist_784', version=1)
        X = mnist.data.values.reshape(-1,28,28)
        y = mnist.target.astype(int).values
        X_train, _, y_train, _ = train_test_split(X, y, test_size=0.2, random_state=42)
        X_train, y_train = X_train[:args.train_samples], y_train[:args.train_samples]

        
        seqs = extract_features(X_train)
        os.makedirs(args.model_dir, exist_ok=True)

        
        for d in range(10):
            digit_seqs = [s for s, lab in zip(seqs, y_train) if lab==d]
            lengths = [len(s) for s in digit_seqs]
            concat = np.vstack(digit_seqs)
            model = hmm.GaussianHMM(n_components=args.states, covariance_type='diag', n_iter=args.iter)
            model.fit(concat, lengths)
            path = os.path.join(args.model_dir, f'model_{d}.pkl')
            with open(path,'wb') as f:
                pickle.dump(model, f)
            print(f"Saved model for digit {d} -> {path}")
    else:
        
        models = {}
        for d in range(10):
            path = os.path.join(args.model_dir, f'model_{d}.pkl')
            with open(path,'rb') as f:
                models[d] = pickle.load(f)

        if args.test_custom:
            if not os.path.exists(args.test_custom):
                print("File not found:", args.test_custom)
            else:
                img = Image.open(args.test_custom).convert('L').resize((28,28))
                seq = ((np.array(img)>128).astype(int).sum(axis=0)/28).reshape(-1,1)
                print(f"[DEBUG] Loaded custom image: {args.test_custom}")
                print(f"[DEBUG] Sequence shape: {seq.shape}")

                scores = {}
                for d in models:
                    try:
                        score = models[d].score(seq)
                        scores[d] = score
                        print(f"Digit {d} score: {score:.4f}")
                    except Exception as e:
                        print(f"Digit {d} failed: {e}")

                if scores:
                    pred = max(scores, key=scores.get)
                    print(f"Custom image {args.test_custom} -> Predicted: {pred}")
                else:
                    print("[ERROR] No valid model scores computed.")

        if args.test_folder:
            if not os.path.isdir(args.test_folder):
                print(f"Folder not found: {args.test_folder}")
            else:
                files = [f for f in os.listdir(args.test_folder) if f.lower().endswith(('.png','.jpg','.jpeg'))]
                for fn in files:
                    fp = os.path.join(args.test_folder, fn)
                    img = Image.open(fp).convert('L').resize((28,28))
                    seq = ((np.array(img)>128).astype(int).sum(axis=0)/28).reshape(-1,1)
                    pred = max({d:models[d].score(seq) for d in models}, key=lambda x:models[x].score(seq))
                    print(f"{fn} -> Predicted: {pred}")
